ecosystem/indexer-grpc/indexer-grpc-file-store-backfiller/src/processor.rs

```diff
@@ -0,0 +1,265 @@
+                    }
+                }
+            });
+            tasks.push(task);
+        }
+        let progress_file_path = self.progress_file_path.clone();
+        let task = tokio::spawn(async move {
+            loop {
+                tokio::time::sleep(Duration::from_millis(5000)).await;
+                {
+                    {
+                        let mut finished_starting_versions =
+                            finished_starting_versions.lock().await;
+                        let mut need_to_update = false;
+                        loop {
+                            if finished_starting_versions.contains(&next_version_to_process) {
+                                finished_starting_versions.remove(&next_version_to_process);
+                                next_version_to_process += 1000;
+                                need_to_update = true;
+                            } else {
+                                break;
+                            }
+                        }
+                        if !need_to_update {
+                            continue;
+                        }
+                    }
+                    // Update the progress file.
+                    let progress_file = ProgressFile {
+                        version: next_version_to_process,
+                    };
+                    let bytes = serde_json::to_vec(&progress_file)
+                        .context("Failed to serialize progress file")?;
+                    std::fs::write(&progress_file_path, &bytes)
+                        .context("Failed to write progress file")?;
+                    tracing::info!(
+                        "Progress file updated to version {}",
+                        next_version_to_process
+                    );
+                    if let Some(ending_version) = ending_version {
+                        if ending_version <= next_version_to_process {
+                            // Backfill is done.
+                            std::process::exit(0);
+                        }
+                    }
+                }
+            }
+        });
+        tasks.push(task);
+
+        // Clone the sender to extend its lifetime.
+        let sender = sender.clone();
+        // Start the stream.
+        loop {
+            let item = grpc_stream.next().await;
+            let item = item.unwrap();
+            let response = match item {
+                Ok(response) => response,
+                Err(e) => {
+                    tracing::error!("Failed to get response: {:?}", e);
+                    panic!();
+                },
+            };
+
+            let resp = response.response.unwrap();
+            match resp {
+                Response::Data(txns) => {
+                    let transactions = txns.transactions;
+                    for txn in transactions {
+                        let version = txn.version;
+                        transactions_buffer.insert(version, txn);
```

Comment:
Why do we need this transactions_buffer?

Reply:
Oh this is to ensure each transactions batch has size 1000. Let's add this in a comment. Overall the flow makes sense, we just need more comments explaining the code.

Reply:
done.